dotCMS/src/main/java/com/dotcms/vanityurl/cache/VanityUrlCacheImpl.java

```diff
@@ -0,0 +1,150 @@
+ * @since June 24, 2017
+ */
+public class VanityUrlCacheImpl extends VanityUrlCache {
+
+    private DotCacheAdministrator cache;
+
+    private static final String VANITY_URL_SITE_GROUP = "VanityURLSiteCache";
+    private static final String VANITY_URL_DIRECT_GROUP = "VanityURLDirectCache";
+
+
+
+    public VanityUrlCacheImpl() {
+        cache = CacheLocator.getCacheAdministrator();
+    }
+
+    @Override
+    public String getPrimaryGroup() {
+
+        return VANITY_URL_SITE_GROUP;
+    }
+
+    @Override
+    public String[] getGroups() {
+
+        return new String[] {VANITY_URL_SITE_GROUP, VANITY_URL_DIRECT_GROUP};
+    }
+
+    @Override
+    public void clearCache() {
+        cache.flushGroup(VANITY_URL_SITE_GROUP);
+        cache.flushGroup(VANITY_URL_DIRECT_GROUP);
+
+    }
+
+    @Override
+    public void remove(final Contentlet vanityURL) {
+        if (vanityURL == null || !vanityURL.isVanityUrl()) {
+            return;
+        }
+        Host host = Try.of(() -> APILocator.getHostAPI().find(vanityURL.getHost(), APILocator.systemUser(), false)).getOrNull();
+        Language lang = Try.of(() -> APILocator.getLanguageAPI().getLanguage(vanityURL.getLanguageId())).getOrNull();
+        if (host == null || lang == null) {
+            return;
+        }
+        remove(host, lang);
+
+
+    }
+
+
+    @Override
+    public void remove(final Host vanityHost) {
+        if (vanityHost == null) {
+            return;
+        }
+
+        final List<Language> langs = APILocator.getLanguageAPI().getLanguages();
+
+        for (Language lang : langs) {
+            remove(vanityHost, lang);
+        }
+
+    }
+
+
+    @Override
+    public void remove(final Host vanityHost, final Language lang) {
+        if (vanityHost == null || lang == null) {
+            return;
+        }
+
+        cache.remove(key(vanityHost, lang), VANITY_URL_SITE_GROUP);
+        cache.flushGroup(VANITY_URL_DIRECT_GROUP);
+
+    }
+
+
+    @Override
+    public void putSiteMappings(final Host host, final Language lang, final List<CachedVanityUrl> vanityURLs) {
+        if (host == null || host.getIdentifier() == null || lang == null || vanityURLs == null) {
+            return;
+        }
+        cache.put(key(host, lang), vanityURLs, VANITY_URL_SITE_GROUP);
+        cache.flushGroup(VANITY_URL_DIRECT_GROUP);
+    }
+
+    @Override
+    public List<CachedVanityUrl> getSiteMappings(final Host host, final Language lang) {
+        if (host == null || lang == null || host.getIdentifier() == null) {
+            return null;
+        }
+        final String key = key(host, lang);
+
+        return (List<CachedVanityUrl>) cache.getNoThrow(key, VANITY_URL_SITE_GROUP);
+
+    }
+
+    @Override
+    public Optional<CachedVanityUrl> getDirectMapping(final String url, final Host host, final Language lang) {
+        Optional<CachedVanityUrl> cachedVanity = (Optional<CachedVanityUrl>) cache.getNoThrow(key(host, lang, url), VANITY_URL_DIRECT_GROUP);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Local variable 'cachedVanity' could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5434315)